services/spar/src/Spar/App.hs

```diff
@@ -134,46 +163,59 @@ toLevel = \case
   SAML.Debug -> Log.Debug
   SAML.Trace -> Log.Trace
 
-instance SPStoreID AuthnRequest Spar where
+instance SPStoreID AuthnRequest (Spar r) where
   storeID i r = wrapMonadClientWithEnv $ Data.storeAReqID i r
   unStoreID r = wrapMonadClient $ Data.unStoreAReqID r
   isAliveID r = wrapMonadClient $ Data.isAliveAReqID r
 
-instance SPStoreID Assertion Spar where
+instance SPStoreID Assertion (Spar r) where
   storeID i r = wrapMonadClientWithEnv $ Data.storeAssID i r
   unStoreID r = wrapMonadClient $ Data.unStoreAssID r
   isAliveID r = wrapMonadClient $ Data.isAliveAssID r
 
-instance SPStoreIdP SparError Spar where
-  type IdPConfigExtra Spar = WireIdP
+instance SPStoreIdP SparError (Spar r) where
+  type IdPConfigExtra (Spar r) = WireIdP
 
-  storeIdPConfig :: IdP -> Spar ()
+  storeIdPConfig :: IdP -> Spar r ()
   storeIdPConfig idp = wrapMonadClient $ Data.storeIdPConfig idp
 
-  getIdPConfig :: IdPId -> Spar IdP
+  getIdPConfig :: IdPId -> Spar r IdP
   getIdPConfig = (>>= maybe (throwSpar SparIdPNotFound) pure) . wrapMonadClientWithEnv . Data.getIdPConfig
 
-  getIdPConfigByIssuer :: Issuer -> Spar IdP
+  getIdPConfigByIssuer :: Issuer -> Spar r IdP
   getIdPConfigByIssuer = (>>= maybe (throwSpar SparIdPNotFound) pure) . wrapMonadClientWithEnv . Data.getIdPConfigByIssuer
 
 -- | 'wrapMonadClient' with an 'Env' in a 'ReaderT', and exceptions. If you
 -- don't need either of those, 'wrapMonadClient' will suffice.
-wrapMonadClientWithEnv :: forall a. ReaderT Data.Env (ExceptT TTLError Cas.Client) a -> Spar a
+wrapMonadClientWithEnv :: forall r a. ReaderT Data.Env (ExceptT TTLError Cas.Client) a -> Spar r a
 wrapMonadClientWithEnv action = do
   denv <- Data.mkEnv <$> (sparCtxOpts <$> ask) <*> (fromTime <$> getNow)
   either (throwSpar . SparCassandraTTLError) pure =<< wrapMonadClient (runExceptT $ action `runReaderT` denv)
 
+instance Member (Final IO) r => Catch.MonadThrow (Sem r) where
+  throwM = embedFinal . Catch.throwM @IO
+
+instance Member (Final IO) r => Catch.MonadCatch (Sem r) where
```

Comment:
These orphan instances are only used for the call to `Catch.catch` on line 212. Is it worth just inlining them there rather than defining the orphans?

Reply:
I suppose that will chip away some of the work of making this an effect as well?  In that case: yes, pls inline; otherwise let's wait for this to be refactored away.